packages/hierarchy-builder/src/test/internal/operators/grouping/AutoExpand.test.ts

```diff
@@ -0,0 +1,352 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { expect } from "chai";
+import { HierarchyNode } from "../../../../hierarchy-builder/HierarchyNode";
+import { assignAutoExpand } from "../../../../hierarchy-builder/internal/operators/grouping/AutoExpand";
+import { createTestNode } from "../../../Utils";
+
+describe("AutoExpand", () => {
```

Comment:
There are 3 sets of nearly identical tests. Maybe we could have something like this to avoid repeating the code?

```ts
[
  [
    "Base class grouping", 
    {
      label: "Base Class",
      key: {
        type: "class-grouping",
        class: {
          name: "TestSchema:BaseClass",
          label: "Base Class",
        },
      },
      children: [
        createTestNode({
          key: { type: "instances", instanceKeys: [{ className: "TestSchema:A", id: "0x1" }] },
          label: "1",
          params: { grouping: { byBaseClasses: { fullClassNames: ["TestSchema:BaseClass"], autoExpand: "single-child" } } },
        }),
      ],
    },
  ],
  [
    "Class grouping",
    { ... },
  ],
  [
    "Label grouping",
    { ... },
  ],
].forEach(([scenario, groupingNode]) => {
  describe(scenario, () => {
    it("sets autoExpand to true when grouping node has one node child and it has autoExpand set to 'single-child'", async () => {
      const nodes: HierarchyNode[] = [groupingNode];
      const result = assignAutoExpand({ grouped: nodes, ungrouped: [], groupingType: "base-class" });
      expect(result.ungrouped).to.deep.eq([]);
      expect(HierarchyNode.isGroupingNode(result.grouped[0])).to.eq(true);
      expect(result.grouped[0].children).to.deep.eq(nodes[0].children);
      expect(result.grouped[0].autoExpand).to.eq(true);
    });
    // ... and other tests here
  });
});
```

Reply:
Changed to this:
```
 const baseClassGroupingParams = {
    testName: "Base class grouping",
    label: "Base Class",
    key: {
      type: "class-grouping",
      class: {
        name: "TestSchema:BaseClass",
        label: "Base Class",
      },
    },
    groupingType: "base-class",
  };
  ...
  describe("sets autoExpand to true when grouping node has one node child and it has autoExpand set to 'single-child'", () => {
    [
      {
        testParams: baseClassGroupingParams,
        processingParams: { grouping: { byBaseClasses: { fullClassNames: ["TestSchema:BaseClass"], autoExpand: "single-child" } } },
      },
      ...
    ].forEach(({ testParams, processingParams }) => {
      it(testParams.testName, async () => {
        const nodes: ProcessedInstancesGroupingHierarchyNode[] = [
          createTestProcessedGroupingNode({
            label: testParams.label,
            key: testParams.key as GroupingNodeKey,
            children: [
              createTestProcessedInstanceNode({
                key: { type: "instances", instanceKeys: [{ className: "TestSchema:A", id: "0x1" }] },
                label: "1",
                processingParams: processingParams as InstanceHierarchyNodeProcessingParams,
              }),
            ],
          }),
        ];
        const result = assignAutoExpand({ grouped: nodes, ungrouped: [], groupingType: testParams.groupingType as GroupingType });
        expect(result.ungrouped).to.deep.eq([]);
        expect(HierarchyNode.isGroupingNode(result.grouped[0])).to.eq(true);
        expect(result.grouped[0].children).to.deep.eq(nodes[0].children);
        expect(result.grouped[0].autoExpand).to.eq(true);
      });
    });
  });
// ... and other tests here
```

Reply:
1. I don't think you need to parametrize the grouping node label. Could just use "Test label" or whatever.
2. I'd rename the `key` parameter name to `groupingNodeKey` to make it easier to understand what it means.
3. Could we move the extra `processingParams` parameters to `testParams`? I think each grouping params could contain a function like 
  ```ts
  {
    createGroupedNodeProcessingParams: (autoExpand: AutoExpand | undefined) => InstanceHierarchyNodeProcessingParams;
  }
  ```
  And then each test could call it to create processing params for grouped nodes.
4. `describe` should describe what you're testing (situation or API) and `it` should describe situation and read like "it does something". This is not the case now - I'd like the structure to be like this:
  ```ts
  [baseClassGroupingParams, classGroupingParams, labelGroupingParams].forEach(({ testName }) => {
    describe(testName, () => {
      it("sets autoExpand to true when grouping node has one node child and it has autoExpand set to 'single-child'", () => {
        // test...
      });

      it("sets autoExpand to true when some child nodes have autoExpand set to 'always'", () => {
        // test...
      });
    });
  });
  ```
  Any reason we can't achieve this?

Reply:
Changed.